Update sampler_t.adoc

JanSellner opened the pull request:
Added examples for the different addressing modes.

- Inspired by OpenCV's [border types](http://docs.opencv.org/master/d2/de8/group__core__array.html#ga209f2f4869e304c82d07739337eae7c5).
- I think it would be great to add to each man page links to existing blog articles or other resources about the topics (e. g. Intel's articles about OpenCL 2.0 features). I started here by introducing a reference section.
- Examples are mainly derived from a small test program (see references). Not 100 % sure, but they seem reasonable to me.

BTW: great idea of launching opencl.org :-)

Commit:
Update sampler_t.adoc

Added examples for the different addressing modes.

jszuppe commented on sampler_t.adoc:
```diff
@@ -46,7 +46,7 @@ The sampler fields are described in the table below:
 |
 
-Specifies whether the `x`, `y` and `z` coordinates are passed in as normalized or unnormalized values.
+Specifies whether the `x`, `y` and `z` coordinates are passed in as normalized ([0,1[) or unnormalized ({0, size-1}) values.
```
Shouldn't it be `[0;1]`? Why do you use `[]` for the first range and `{}` for the 2nd range?

JanSellner replied:
Wow, that was fast :-)

This was the result from [my testing](https://www.milania.de/blog/Addressing_mode_of_sampler_objects_for_image_types_in_OpenCL_reviewed). Relevant lines from the AMD output for accessing an 1D image with values 1,...,10:
```
 0.90000 |              10 |     10 |            10 |    10
 1.00000 |              10 |      1 |            10 |     0
```
This led me to the conclusion that 1.0 already gives an out of border value and therefore the right side is not included (hence `[`).

`[]` because by using normalised coordinates all continuous values between 0 and 1 are possible and `{}` because only the discrete values `{0,1,2,...,size-1}` are valid.

jszuppe replied:
>This led me to the conclusion that 1.0 already gives an out of border value and therefore the right side is not included (hence `[`).

Oh, I did not know that `[` means that the right side is not included. I always use ')' to express that. In that case it's ok. 

>`[]` because by using normalised coordinates all continuous values between 0 and 1 are possible and `{}` because only the discrete values `{0,1,2,...,size-1}` are valid.

Right. I would replace  `{0, size-1}`  with `{0, ..., size-1}`, so it's easier to understand. Are you sure they must be integers?

I was able to find that it works the other way around: if you use integer coordinates sampler must be set in `CLK_NORMALIZED_COORDS_FALSE` mode.

>The read_imagef calls that take integer coordinates must use a sampler with filter mode set to
CLK_FILTER_NEAREST, normalized coordinates set to CLK_NORMALIZED_COORDS_FALSE and addressing mode set to CLK_ADDRESS_CLAMP_TO_EDGE, CLK_ADDRESS_CLAMP or CLK_ADDRESS_NONE; otherwise the values returned are undefined

On the other hand, if you don't use normalized coords, then you probably want to use integer coordinates (`{0, ..., size-1}`).

JanSellner replied:
Thanks, I forgot the points.

Actually, I don't know if `{}` also implies integer coordinates, `{0.0, 1.0, 2.0, ...}` would also be possible, I guess. Anyway, I think it is clear know.

jszuppe replied:
Wouldn't 10.5 also be a correct coordinate? If yes, then it should be `[0, size - 1]`.

JanSellner replied:
I just tested by accessing the coordinates `{-4.5, -5.0, ..., 15.0, 15.5}` with `CLK_NORMALIZED_COORDS_FALSE` and this was the result:
```
Platforms
        0: Intel(R) OpenCL - OpenCL 1.2
        1: Experimental OpenCL 2.1 CPU Only Platform - OpenCL 2.1
        2: AMD Accelerated Parallel Processing - OpenCL 2.0 AMD-APP (2348.3)
Choose plattform: 2
Devices on plattform AMD Accelerated Parallel Processing
        0: Hawaii - OpenCL 2.0 AMD-APP (2348.3)
        1: Intel(R) Core(TM) i7-5820K CPU @ 3.30GHz - OpenCL 1.2 AMD-APP (2348.3)
Choose device: 0
  coords | CLAMP_TO_EDGE | CLAMP
---------|---------------|---------
-4.50000 |       1.00000 | 0.00000
-3.50000 |       1.00000 | 0.00000
-2.50000 |       1.00000 | 0.00000
-1.50000 |       1.00000 | 0.00000
-0.50000 |       1.00000 | 0.00000
 0.50000 |       1.00000 | 1.00000
 1.50000 |       2.00000 | 2.00000
 2.50000 |       3.00000 | 3.00000
 3.50000 |       4.00000 | 4.00000
 4.50000 |       5.00000 | 5.00000
 5.50000 |       6.00000 | 6.00000
 6.50000 |       7.00000 | 7.00000
 7.50000 |       8.00000 | 8.00000
 8.50000 |       9.00000 | 9.00000
 9.50000 |      10.00000 | 10.00000
10.50000 |      10.00000 | 0.00000
11.50000 |      10.00000 | 0.00000
12.50000 |      10.00000 | 0.00000
13.50000 |      10.00000 | 0.00000
14.50000 |      10.00000 | 0.00000
15.50000 |      10.00000 | 0.00000
```
So, it seems to be that even with unnormalised coordinates it is possible to access every value. This would mean that the actual possible range is `]-\infty,\infty[`, but that is not really helpful in my opinion^^.

I think I would keep `{0, ..., size-1}`, since this is the range of reasonable coordinates. But maybe an additional entry in the notes section describing that it is in both cases possible to use coordinates outside of the defined range?

jszuppe replied:
Oh, I wasn't talking about out of range coordinates. I wanted to say that since floating point coordinates work with unnormalized values, we should just say that unnormalized values are in `[0; size - 1]` range (not only integers), or maybe `[0; dimension_size - 1]`. If developer uses function with int coordinates, it's obvious she/he needs to use `intN`.

JanSellner replied:
Ah, I see. Fair point. I changed it to `[0, dimension_size - 1]`.

jszuppe commented:
Thanks! Those're exactly the changes we are looking for.

Commit:
Update sampler_t.adoc

Clarification for the coordinate ranges.

Commit:
Update sampler_t.adoc

Changed brackets for unnormalised coordinates since floating coordinates can be used as well

jszuppe commented:
Thanks! Merged!
